pyo3-macros-backend/src/utils.rs

```diff
@@ -67,14 +68,58 @@ pub fn option_type_argument(ty: &syn::Type) -> Option<&syn::Type> {
     None
 }
 
+// TODO: Replace usage of this by [`syn::LitCStr`] when on MSRV 1.77
+#[derive(Clone)]
+pub struct LitCStr {
+    lit: CString,
+    span: Span,
+    #[allow(dead_code)]
+    ctx: Ctx,
+}
+
+impl LitCStr {
+    pub fn new(lit: CString, span: Span, ctx: Ctx) -> Self {
+        Self { lit, span, ctx }
+    }
+
+    pub fn empty(ctx: Ctx) -> Self {
+        Self {
+            lit: CString::new("").unwrap(),
+            span: Span::call_site(),
+            ctx,
+        }
+    }
+}
+
+impl quote::ToTokens for LitCStr {
+    #[cfg(c_str_lit)]
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        syn::LitCStr::new(&self.lit, self.span).to_tokens(tokens);
+    }
+
+    #[cfg(not(c_str_lit))]
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let Ctx { pyo3_path, .. } = &self.ctx;
+        let lit = self.lit.to_str().unwrap();
+        tokens.extend(quote::quote_spanned!(self.span => #pyo3_path::ffi::c_str!(#lit)));
+    }
+}
```

Comment:
```suggestion
impl quote::ToTokens for LitCStr {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        if cfg!(c_str_lit) {
             syn::LitCStr::new(&self.lit, self.span).to_tokens(tokens);
        } else {
             let Ctx { pyo3_path, .. } = &self.ctx;
             let lit = self.lit.to_str().unwrap();
             tokens.extend(quote::quote_spanned!(self.span => #pyo3_path::ffi::c_str!(#lit)));
        }
    }
}
```

Then you can also remove the dead_code allow used earlier ;)